core/src/main/java/hudson/Util.java

```diff
@@ -507,39 +511,23 @@ public static boolean isSymlink(@Nonnull File file) throws IOException {
          *  both as symlinks, let's use that function and always call it before calling down to the
          *  NIO2 API.
          *
+         *  JENKINS-39179, JENKINS-36088.
+         *  JNA can cause deadlocks, so we avoid it. WindowsFileAttributes.isOther() includes devices,
+         *  but reading the attributes of a device with NIO fails or returns false for isOther(). Named
+         *  pipes are not devices, and return false for isOther().
          */
-        if (Functions.isWindows()) {
-            try {
-                return Kernel32Utils.isJunctionOrSymlink(file);
-            } catch (UnsupportedOperationException | LinkageError e) {
-                // fall through
-            }
-        }
-        Boolean r = isSymlinkJava7(file);
-        if (r != null) {
-            return r;
-        }
-        String name = file.getName();
-        if (name.equals(".") || name.equals(".."))
-            return false;
-
-        File fileInCanonicalParent;
-        File parentDir = file.getParentFile();
-        if ( parentDir == null ) {
-            fileInCanonicalParent = file;
-        } else {
-            fileInCanonicalParent = new File( parentDir.getCanonicalPath(), name );
-        }
-        return !fileInCanonicalParent.getCanonicalFile().equals( fileInCanonicalParent.getAbsoluteFile() );
-    }
-
-    @SuppressFBWarnings("NP_BOOLEAN_RETURN_NULL")
-    private static Boolean isSymlinkJava7(@Nonnull File file) throws IOException {
         try {
             Path path = file.toPath();
-            return Files.isSymbolicLink(path);
-        } catch (Exception x) {
-            throw (IOException) new IOException(x.toString()).initCause(x);
+            BasicFileAttributes attrs = Files.readAttributes(path, BasicFileAttributes.class, LinkOption.NOFOLLOW_LINKS);
+            if (attrs.isSymbolicLink()) {
+                return true;
+            } else if (attrs instanceof DosFileAttributes) {
+                return attrs.isOther();
```

Comment:
We could use reflection to call `WindowsFileAttributes#isReparsePoint` instead to avoid including devices, but see the original comment where I was unable to read the attributes of a device that returned true for `isOther`.

Reply:
allthough you have a detailed comment in the javadoc above - it would be good to leave a small comment here as well!